presto-hudi/src/main/java/com/facebook/presto/hudi/HudiMetadata.java

```diff
@@ -0,0 +1,272 @@
+    }
+
+    @Override
+    public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTableHandle tableHandle, ColumnHandle columnHandle)
+    {
+        return ((HudiColumnHandle) columnHandle).toColumnMetadata(typeManager);
+    }
+
+    @Override
+    public Map<SchemaTableName, List<ColumnMetadata>> listTableColumns(ConnectorSession session, SchemaTablePrefix prefix)
+    {
+        List<SchemaTableName> tables = prefix.getTableName() != null ? singletonList(prefix.toSchemaTableName()) : listTables(session, Optional.of(prefix.getSchemaName()));
+
+        ImmutableMap.Builder<SchemaTableName, List<ColumnMetadata>> columns = ImmutableMap.builder();
+        for (SchemaTableName table : tables) {
+            try {
+                columns.put(table, getTableMetadata(session, table).getColumns());
+            }
+            catch (TableNotFoundException e) {
+                log.warn(String.format("table disappeared during listing operation: %s", e.getMessage()));
+            }
+        }
+        return columns.build();
+    }
+
+    public ExtendedHiveMetastore getMetastore()
+    {
+        return metastore;
+    }
+
+    private Table getTable(ConnectorSession connectorSession, ConnectorTableHandle tableHandle)
+    {
+        MetastoreContext metastoreContext = toMetastoreContext(connectorSession);
+        HudiTableHandle handle = (HudiTableHandle) tableHandle;
+        Optional<Table> table = metastore.getTable(metastoreContext, handle.getSchemaName(), handle.getTableName());
+        checkArgument(table.isPresent());
+        return table.get();
+    }
+
+    private ConnectorTableMetadata getTableMetadata(ConnectorSession session, SchemaTableName tableName)
+    {
+        Table table = metastore.getTable(
+                toMetastoreContext(session),
+                tableName.getSchemaName(),
+                tableName.getTableName()).orElseThrow(() -> new TableNotFoundException(tableName));
+
+        List<ColumnMetadata> columnMetadatas = allColumnHandles(table)
+                .map(columnHandle -> columnHandle.toColumnMetadata(typeManager))
+                .collect(toList());
+        return new ConnectorTableMetadata(tableName, columnMetadatas);
+    }
+
+    private Stream<HudiColumnHandle> allColumnHandles(Table table)
+    {
+        return Stream.concat(getDataColumnHandles(table).stream(), getPartitionColumnHandles(table).stream());
+    }
+
+    private List<HudiColumnHandle> getDataColumnHandles(Table table)
+    {
+        return fromDataColumns(table.getDataColumns());
+    }
+
+    private List<HudiColumnHandle> getPartitionColumnHandles(Table table)
+    {
+        return fromPartitionColumns(table.getPartitionColumns());
+    }
+
+    static List<HudiColumnHandle> fromPartitionColumns(List<Column> partitionColumns)
+    {
+        ImmutableList.Builder<HudiColumnHandle> builder = ImmutableList.builder();
```

Comment:
nit: create builderWithSize to avoid reallocations.

Reply:
Fixed.